src/hotspot/share/opto/vtransform.hpp

```diff
@@ -0,0 +1,516 @@
+  virtual VTransformApplyResult apply(const VLoopAnalyzer& vloop_analyzer,
+                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const override;
+  NOT_PRODUCT(virtual const char* name() const override { return "Replicate"; };)
+  NOT_PRODUCT(virtual void print_spec() const override;)
+};
+
+// Transform introduces a scalar ConvI2LNode that was not previously in the C2 graph.
+class VTransformConvI2LNode : public VTransformNode {
+public:
+  VTransformConvI2LNode(VTransform& vtransform) : VTransformNode(vtransform, 2) {}
+  virtual VTransformApplyResult apply(const VLoopAnalyzer& vloop_analyzer,
+                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const override;
+  NOT_PRODUCT(virtual const char* name() const override { return "ConvI2L"; };)
+};
+
+// Transform introduces a shift-count node, that truncates the shift count for a vector shift.
+class VTransformShiftCountNode : public VTransformNode {
+private:
+  int _vlen;
+  const BasicType _element_bt;
+  juint _mask;
+  int _shift_opcode;
+public:
+  VTransformShiftCountNode(VTransform& vtransform, int vlen, BasicType element_bt, juint mask, int shift_opcode) :
+    VTransformNode(vtransform, 2), _vlen(vlen), _element_bt(element_bt), _mask(mask), _shift_opcode(shift_opcode) {}
+  virtual VTransformApplyResult apply(const VLoopAnalyzer& vloop_analyzer,
+                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const override;
+  NOT_PRODUCT(virtual const char* name() const override { return "ShiftCount"; };)
+  NOT_PRODUCT(virtual void print_spec() const override;)
+};
+
+// Transform introduces a PopulateIndex node: [phi, phi+1, phi+2, phi+3, ...].
+class VTransformPopulateIndexNode : public VTransformNode {
+private:
+  int _vlen;
+  const BasicType _element_bt;
+public:
+  VTransformPopulateIndexNode(VTransform& vtransform, int vlen, const BasicType element_bt) :
+    VTransformNode(vtransform, 2), _vlen(vlen), _element_bt(element_bt) {}
+  virtual VTransformApplyResult apply(const VLoopAnalyzer& vloop_analyzer,
+                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const override;
+  NOT_PRODUCT(virtual const char* name() const override { return "PopulateIndex"; };)
+  NOT_PRODUCT(virtual void print_spec() const override;)
+};
+
+// Base class for all vector vtnodes.
+class VTransformVectorNode : public VTransformNode {
+private:
+  GrowableArray<Node*> _nodes;
+public:
+  VTransformVectorNode(VTransform& vtransform, const uint req, const uint number_of_nodes) :
+    VTransformNode(vtransform, req), _nodes(vtransform.arena(), number_of_nodes, number_of_nodes, nullptr) {}
+
+  void set_nodes(const Node_List* pack) {
+    for (uint k = 0; k < pack->size(); k++) {
+      _nodes.at_put(k, pack->at(k));
+    }
+  }
+
+  const GrowableArray<Node*> nodes() const { return _nodes; }
```

Comment:
Oh wow, that is a good catch!
I guess under the hood this did not make a difference, since `GrowableArray` does shallow copy, and while we use the copy we would not modify the original. But still, this should be a reference ;)